From 50668ec0710823a7990c46dd40a8fc4431135c24 Mon Sep 17 00:00:00 2001
From: Miguel Vadillo <vadillo@ti.com>
Date: Sat, 12 Mar 2011 18:14:55 -0600
Subject: syslink: ipu_pm: relax syslink fixed constraints

Relax the core constraint requested by ipu when in use.
Also request the constraint when ipu_pm_setup is done and
not until the first message is sent.

Change-Id: Iccefca9b12f3f9bf09138150cc6dfa465305af2e
Signed-off-by: Paul Hunt <hunt@ti.com>
Signed-off-by: Miguel Vadillo <vadillo@ti.com>
---
 drivers/dsp/syslink/ipu_pm/ipu_pm.c |   41 +++++++++++++++++++++-------------
 drivers/dsp/syslink/ipu_pm/ipu_pm.h |    5 ++-
 2 files changed, 28 insertions(+), 18 deletions(-)

Index: kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.c
===================================================================
--- kernel.orig/drivers/dsp/syslink/ipu_pm/ipu_pm.c	2011-04-06 22:17:30.000000000 +0900
+++ kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.c	2011-04-06 22:20:16.000000000 +0900
@@ -1863,7 +1863,7 @@
 	if (params->pm_iva_hd_counter == 0 && params->pm_iss_counter == 0) {
 		pr_debug("Release MPU wakeup latency\n");
 		retval = omap_pm_set_max_mpu_wakeup_lat(&pm_qos_handle,
-						IPU_PM_NO_MPU_LAT_CONSTRAINT);
+						IPU_PM_NO_LAT_CONSTRAINT);
 		if (retval) {
 			pr_err("%s %d Error setting MPU cstr\n"
 							, __func__, __LINE__);
@@ -1978,7 +1978,7 @@
 	if (params->pm_iva_hd_counter == 0 && params->pm_iss_counter == 0) {
 		pr_debug("Release MPU wakeup latency\n");
 		retval = omap_pm_set_max_mpu_wakeup_lat(&pm_qos_handle,
-						IPU_PM_NO_MPU_LAT_CONSTRAINT);
+						IPU_PM_NO_LAT_CONSTRAINT);
 		if (retval) {
 			pr_err("%s %d Error setting MPU cstr\n"
 							, __func__, __LINE__);
@@ -2474,13 +2474,6 @@
 		return 0;
 	}
 
-#ifdef CONFIG_OMAP_PM
-	retval = omap_pm_set_max_sdma_lat(&pm_qos_handle_2,
-						IPU_PM_NO_MPU_LAT_CONSTRAINT);
-	if (retval)
-		pr_info("Unable to remove cstr on IPU\n");
-#endif
-
 	/* Because of the current scheme, we need to check
 	 * if APPM3 is enable and we need to shut it down too
 	 * Sysm3 is the only want sending the hibernate message
@@ -2530,6 +2523,13 @@
 			goto error;
 		handle->rcb_table->state_flag |= SYS_PROC_DOWN;
 
+#ifdef CONFIG_OMAP_PM
+		retval = omap_pm_set_max_sdma_lat(&pm_qos_handle_2,
+						IPU_PM_NO_LAT_CONSTRAINT);
+		if (retval)
+			pr_info("Unable to remove cstr on IPU\n");
+#endif
+
 		/* If there is a message in the mbox restore
 		 * immediately after save.
 		 */
@@ -2606,12 +2606,6 @@
 		cm_write_mod_reg(HW_AUTO,
 				 OMAP4430_CM2_CORE_MOD,
 				 OMAP4_CM_DUCATI_CLKSTCTRL_OFFSET);
-#ifdef CONFIG_OMAP_PM
-		retval = omap_pm_set_max_sdma_lat(&pm_qos_handle_2,
-						IPU_PM_MM_MPU_LAT_CONSTRAINT);
-		if (retval)
-			pr_info("Unable to set cstr on IPU\n");
-#endif
 	}
 
 	/* Check if the M3 was loaded */
@@ -2665,7 +2659,7 @@
 #endif
 #ifdef CONFIG_OMAP_PM
 		retval = omap_pm_set_max_sdma_lat(&pm_qos_handle_2,
-						IPU_PM_MM_MPU_LAT_CONSTRAINT);
+						IPU_PM_MM_SDMA_LAT_CONSTRAINT);
 		if (retval)
 			pr_info("Unable to set cstr on IPU\n");
 #endif
@@ -2793,6 +2787,13 @@
 #endif
 	BLOCKING_INIT_NOTIFIER_HEAD(&ipu_pm_notifier);
 
+#ifdef CONFIG_OMAP_PM
+	retval = omap_pm_set_max_sdma_lat(&pm_qos_handle_2,
+						IPU_PM_MM_SDMA_LAT_CONSTRAINT);
+	if (retval)
+		pr_info("Unable to set cstr on IPU\n");
+#endif
+
 	return retval;
 exit:
 	pr_err("ipu_pm_setup failed! retval = 0x%x", retval);
@@ -3079,6 +3080,14 @@
 exit:
 	if (retval < 0)
 		pr_err("ipu_pm_destroy failed, retval: %x\n", retval);
+
+#ifdef CONFIG_OMAP_PM
+	retval = omap_pm_set_max_sdma_lat(&pm_qos_handle_2,
+						IPU_PM_NO_LAT_CONSTRAINT);
+	if (retval)
+		pr_info("Unable to remove cstr on IPU\n");
+#endif
+
 	return retval;
 }
 EXPORT_SYMBOL(ipu_pm_destroy);
Index: kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.h
===================================================================
--- kernel.orig/drivers/dsp/syslink/ipu_pm/ipu_pm.h	2011-04-05 23:58:12.000000000 +0900
+++ kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.h	2011-04-06 22:21:11.000000000 +0900
@@ -111,7 +111,8 @@
 #define PM_CSTR_BW_MASK		0x00000004
 
 #define IPU_PM_MM_MPU_LAT_CONSTRAINT	400
-#define IPU_PM_NO_MPU_LAT_CONSTRAINT	-1
+#define IPU_PM_MM_SDMA_LAT_CONSTRAINT	400
+#define IPU_PM_NO_LAT_CONSTRAINT	-1
 #define NO_FREQ_CONSTRAINT		0
 #define NO_LAT_CONSTRAINT		-1
 #define NO_BW_CONSTRAINT		-1
